### DISCUSSION TITLE
[CPP-CPD] Clones blocks end when a variable is renamed

### DISCUSSION THREAD

**Discussion by theoleleu** [pmd/pmd#3673] [Q&A] [answered]
Hi,

I'm using CPD in CPP for clone detection.
I would like to detect clones even if some variables are renamed.
As I can see in the tool, the detected block ends when detecting a variable renamed, is it possible to bypass it in order to detect bigger clones while not taking into account variable names.

Thanks in advance :)

**Comment by adangel** [comment, marked as answer]
For Java we have the option [`--ignore-identifiers`](https://pmd.github.io/latest/pmd_userdocs_cpd.html#-ignore-identifiers) which probably does what you want. But that's for Java only.
For C++ that would need to be implemented though.

In Java, we replace the identifier tokens with a constant string (javaToken.kind here is always "JavaParserConstants.IDENTIFIER"):
https://github.com/pmd/pmd/blob/641b0e953d87ebb5e544650f0e4aa3c3336b4b5c/pmd-java/src/main/java/net/sourceforge/pmd/cpd/JavaTokenizer.java#L69-L71

That would need to be implemented in a similar way for C++. In [CPPTokenizer](https://github.com/pmd/pmd/blob/master/pmd-cpp/src/main/java/net/sourceforge/pmd/cpd/CPPTokenizer.java) the method processToken needs to be overwritten. Currently this is what's in the super class: https://github.com/pmd/pmd/blob/641b0e953d87ebb5e544650f0e4aa3c3336b4b5c/pmd-core/src/main/java/net/sourceforge/pmd/cpd/internal/JavaCCTokenizer.java#L27-L29
